<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Comparative methods on networks - Network calibration</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./topic2-phyloANOVA.html" rel="next">
<link href="./topic-overview.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Comparative methods on networks</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="./index.html" aria-current="page">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="http://crsl4.github.io/PhyloNetworks.jl/latest/">
 <span class="menu-text">PhyloNetworks</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/cecileane/networkPCM-workshop"><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">Network calibration</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation docked overflow-auto">
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Home</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./topic-overview.html" class="sidebar-item-text sidebar-link">Workshop overview</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./topic1-netcalibration.html" class="sidebar-item-text sidebar-link active">Network calibration</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./topic2-phyloANOVA.html" class="sidebar-item-text sidebar-link">Phylogenetic regression</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./topic3-asr.html" class="sidebar-item-text sidebar-link">Ancestral state reconstruction</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./topic4-impactgeneflow.html" class="sidebar-item-text sidebar-link">Impact of gene flow</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./topic5-discrete.html" class="sidebar-item-text sidebar-link">Discrete trait</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./topic6-visualization.html" class="sidebar-item-text sidebar-link">Visualization</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./topic7-averagedistances.html" class="sidebar-item-text sidebar-link">Average distances</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#input-network" id="toc-input-network" class="nav-link active" data-scroll-target="#input-network">input network</a></li>
  <li><a href="#input-genetic-distances" id="toc-input-genetic-distances" class="nav-link" data-scroll-target="#input-genetic-distances">input genetic distances</a></li>
  <li><a href="#run-the-calibration" id="toc-run-the-calibration" class="nav-link" data-scroll-target="#run-the-calibration">run the calibration</a></li>
  <li><a href="#normalize-the-network-height" id="toc-normalize-the-network-height" class="nav-link" data-scroll-target="#normalize-the-network-height">normalize the network height</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/cecileane/networkPCM-workshop/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">Network calibration</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p>If not done during the session before, load necessary packages</p>
<div class="cell" data-execution_count="1">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">CSV               </span><span class="co"># read/write CSV and similar formats</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">DataFrames        </span><span class="co"># versatile tabular data format</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">PhyloNetworks     </span><span class="co"># includes many utilities</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">PhyloPlots        </span><span class="co"># for plotting networks: via R</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">RCall             </span><span class="co"># run R within Julia</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">StatsBase</span>, <span class="bu">StatsModels</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="input-network" class="level2">
<h2 class="anchored" data-anchor-id="input-network">input network</h2>
<p>We first read the network, as shown in the overview section. In that section, the <a href="./topic-overview.html#fig-0-netsnaq">figure</a> shows that the network needs to be rooted correctely, is missing some edge lengths, and that current edge lengths are in coalescent units. These units are expected to correlate poorly with time, because the number of generations per year can vary significantly between species, and the effective population size can also vary greatly across time species (e.g.&nbsp;bottlenecks at speciation or reticulations) and across species.</p>
<p>Below, the calibration method will ignore these branch lengths in coalescent units and use genetic distances instead, in average number of substitutions/site (averaged across genes).</p>
<div class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>net_snaq <span class="op">=</span> <span class="fu">readTopology</span>(<span class="st">"data/polemonium_network_fromSNaQ.phy"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">tipLabels</span>(net_snaq)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="3">
<pre><code>17-element Vector{String}:
 "I23917_JR124_Polemonium_viscosum"
 "I23927_JR5496_Polemonium_elusum"
 "I26284_JRRS15_Polemonium_eddyense"
 "I23919_JR107_Polemonium_chartaceum"
 "I26282_JR137_Polemonium_micranthum"
 "I23920_JRRS17_Polemonium_occidentale_occidentale"
 "I20029_JR11_Polemonium_reptans"
 "I23907_JR95_Polemonium_pectinatum"
 "I26285_JR136_Polemonium_delicatum"
 "I23883_JRRS16_Polemonium_pulcherimum_shastense"
 "I20028_JR8_Polemonium_elegans"
 "I23905_JRRS10_Polemonium_carneum"
 "I23912_JR112_Polemonium_pauciflorum"
 "I23921_JR109_Polemonium_viscosum"
 "I23914_JR098_Polemonium_brandegeii"
 "I23913_JR040_Polemonium_molle"
 "I20026_JR32_Polemonium_filicinum"</code></pre>
</div>
</div>
<p>Before moving on, let’s simplify taxon names. We have a file that gives the correspondence between tip names in the network to “morph” names in the trait data:</p>
<div class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>namemap <span class="op">=</span> CSV.<span class="fu">read</span>(<span class="st">"data/tip_to_morph.csv"</span>, DataFrame);</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">first</span>(namemap, <span class="fl">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="4">

<div><div style="float: left;"><span>5×2 DataFrame</span></div><div style="clear: both;"></div></div><div class="data-frame" style="overflow-x: scroll;"><table class="data-frame table table-sm table-striped" style="margin-bottom: 6px;"><thead><tr class="header"><th class="rowNumber" style="font-weight: bold; text-align: right;">Row</th><th style="text-align: left;">tip</th><th style="text-align: left;">morph</th></tr><tr class="subheader headerLastRow"><th class="rowNumber" style="font-weight: bold; text-align: right;"></th><th title="String" style="text-align: left;">String</th><th title="String31" style="text-align: left;">String31</th></tr></thead><tbody><tr><td class="rowNumber" style="font-weight: bold; text-align: right;">1</td><td style="text-align: left;">I26282_JR137_Polemonium_micranthum</td><td style="text-align: left;">micranthum</td></tr><tr><td class="rowNumber" style="font-weight: bold; text-align: right;">2</td><td style="text-align: left;">I23899_JR133_Polemonium_pulcherrimum_lindleyi</td><td style="text-align: left;">pulcherrimum_lindleyi</td></tr><tr><td class="rowNumber" style="font-weight: bold; text-align: right;">3</td><td style="text-align: left;">I23916_JR128_Polemonium_pulcherrimum_pulcherrimum</td><td style="text-align: left;">pulcherrimum_pulcherrimum</td></tr><tr><td class="rowNumber" style="font-weight: bold; text-align: right;">4</td><td style="text-align: left;">I23925_JRRS20_Polemonium_pulcherrimum_pulcherrimum</td><td style="text-align: left;">pulcherrimum_pulcherrimum_2</td></tr><tr><td class="rowNumber" style="font-weight: bold; text-align: right;">5</td><td style="text-align: left;">I23883_JRRS16_Polemonium_pulcherimum_shastense</td><td style="text-align: left;">pulcherrimum_shastense</td></tr></tbody></table></div>
</div>
</div>
<p>Let’s build a dictionary to make it easier to do the mapping:</p>
<div class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>tip2morph <span class="op">=</span> <span class="fu">Dict</span>(r[<span class="op">:</span>tip] <span class="op">=&gt;</span> r[<span class="op">:</span>morph] <span class="cf">for</span> r <span class="kw">in</span> <span class="fu">eachrow</span>(namemap))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="5">
<pre><code>Dict{String, String31} with 45 entries:
  "I23906_JRRS22_Polemonium_eximium"            =&gt; "eximium"
  "I23891_JR003_Polemonium_vanbruntiae"         =&gt; "vanbruntiae"
  "I23901_JR091_Polemonium_occidentale_occiden… =&gt; "occidentale_occidentale_2"
  "I23925_JRRS20_Polemonium_pulcherrimum_pulch… =&gt; "pulcherrimum_pulcherrimum_2"
  "I23892_JR038_Polemonium_foliosissimum"       =&gt; "filicinum"
  "I23913_JR040_Polemonium_molle"               =&gt; "foliosissimum_2"
  "I23927_JR5496_Polemonium_elusum"             =&gt; "elusum_2"
  "I23899_JR133_Polemonium_pulcherrimum_lindle… =&gt; "pulcherrimum_lindleyi"
  "I23904_JRRS18_Polemonium_californicum"       =&gt; "californicum_2"
  "I23893_JR122_Ericales_Fouquieriaceae_Fouqui… =&gt; "foliosissimum_3"
  "I26285_JR136_Polemonium_delicatum"           =&gt; "delicatum_2"
  "I23908_JR104_Polemonium_acutiflorum"         =&gt; "acutiflorum"
  "I26283_JR037_Polemonium_viscosum"            =&gt; "viscosum"
  "I23903_JR102_Polemonium_flavum"              =&gt; "flavum"
  "I20027_JR46_Polemonium_alpinum"              =&gt; "albiflorum"
  "I20028_JR8_Polemonium_elegans"               =&gt; "elegans"
  "I23898_JRRS25_Polemonium_viscosum"           =&gt; "aff._viscosum_sp._nov."
  "I23897_JR5148_Polemonium_elusum"             =&gt; "elusum"
  "I23889_JR041_Polemonium_molle"               =&gt; "foliosissimum"
  "I23902_JR101_Polemonium_delicatum"           =&gt; "delicatum"
  "I23883_JRRS16_Polemonium_pulcherimum_shaste… =&gt; "pulcherrimum_shastense"
  "I23905_JRRS10_Polemonium_carneum"            =&gt; "carneum_2"
  "I20030_JRVB3_Polemonium_vanbruntiae"         =&gt; "vanbruntiae_2"
  "I23890_JRRS14_Polemonium_eximium"            =&gt; "eximium_2"
  "I23907_JR95_Polemonium_pectinatum"           =&gt; "pectinatum"
  ⋮                                             =&gt; ⋮</code></pre>
</div>
</div>
<div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> tip <span class="kw">in</span> net_snaq.node</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  tip.leaf <span class="op">||</span> <span class="cf">continue</span> <span class="co"># if not leaf: continue to next iternation (skip next line)</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  tip.name <span class="op">=</span> tip2morph[tip.name]</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># root the network correctly. Now we can use the short names</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="fu">rootatnode!</span>(net_snaq, <span class="st">"micranthum"</span>) <span class="co"># 1 outgroup: micranthum. otherwise use rootonedge!</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the network to see where we should rotate edges to uncross reticulations</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>R<span class="st">"par"</span>(mar<span class="op">=</span>[<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>]); <span class="co"># change default margins to 0</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>res <span class="op">=</span> <span class="fu">plot</span>(net_snaq, shownodenumber<span class="op">=</span><span class="cn">true</span>);</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>res[<span class="fl">1</span><span class="op">:</span><span class="fl">2</span>] <span class="co"># (0.0, 12.1) : default x limits for this network</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> nodenumber <span class="kw">in</span> [<span class="op">-</span><span class="fl">16</span>, <span class="op">-</span><span class="fl">17</span>,<span class="op">-</span><span class="fl">6</span>,<span class="op">-</span><span class="fl">4</span>, <span class="op">-</span><span class="fl">19</span>,<span class="op">-</span><span class="fl">21</span>]</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rotate!</span>(net_snaq, nodenumber)</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(net_snaq, xlim<span class="op">=</span>[<span class="fl">0</span>,<span class="fl">16</span>]); <span class="co"># extend limits to show full taxon names</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-1-netsnaq" class="cell quarto-layout-panel" data-execution_count="5">
<figure class="figure">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-1-netsnaq-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="topic1-netcalibration_files/figure-html/fig-1-netsnaq-output-1.png" class="img-fluid figure-img" data-ref-parent="fig-1-netsnaq"></p>
<p></p><figcaption class="figure-caption">(a) before rotating edges</figcaption><p></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-1-netsnaq-2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="topic1-netcalibration_files/figure-html/fig-1-netsnaq-output-2.png" class="img-fluid figure-img" data-ref-parent="fig-1-netsnaq"></p>
<p></p><figcaption class="figure-caption">(b) after rotating edges</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;1: Polemonium network: rooted</figcaption><p></p>
</figure>
</div>
</div>
</section>
<section id="input-genetic-distances" class="level2">
<h2 class="anchored" data-anchor-id="input-genetic-distances">input genetic distances</h2>
<p>To calibrate the network, we need genetic distances estimated between each pair of taxa. The calibration will find edge lengths in the network to match as best as possible these input pairwise distances between taxa.</p>
<p>These distances were obtained by estimating a gene tree for each gene, rescaling the gene trees to decrease rate variation across genes, extracting the pairwise distances from each rescaled gene tree, then average the pairwise distances across genes (see <a href="./topic7-averagedistances.html">here</a> for how to do this). The results were saved in file <code>data/polemonium_geneticpairwisedistances.csv</code>, which we read below.</p>
<div class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>avD <span class="op">=</span> CSV.<span class="fu">read</span>(<span class="st">"data/polemonium_geneticpairwisedistances.csv"</span>, DataFrame);</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>taxa_long <span class="op">=</span> <span class="fu">names</span>(avD) <span class="co"># column names</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>taxa <span class="op">=</span> [tip2morph[name] for name <span class="kw">in</span> taxa_long]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="7">
<pre><code>17-element Vector{String31}:
 "aff._viscosum_sp._nov._2"
 "elusum_2"
 "eddyense"
 "chartaceum_2"
 "micranthum"
 "occidentale_occidentale"
 "reptans_2"
 "pectinatum"
 "delicatum_2"
 "pulcherrimum_shastense"
 "elegans"
 "carneum_2"
 "pauciflorum"
 "confertum"
 "brandegeei"
 "foliosissimum_2"
 "apachianum"</code></pre>
</div>
</div>
<p>The order in which taxa appear in the matrix is important, which is why we extracted the taxa (ordered) from the column names.</p>
<div class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>avD <span class="op">=</span> <span class="fu">Matrix</span>(avD) <span class="co"># converting to a matrix removes column names</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>avD <span class="op">|&gt;</span> maximum <span class="co"># max distance: ≈ 0.038 substitutions/site, averaged across genes</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="8">
<pre><code>0.038069708207402644</code></pre>
</div>
</div>
</section>
<section id="run-the-calibration" class="level2">
<h2 class="anchored" data-anchor-id="run-the-calibration">run the calibration</h2>
<p>We now have the 2 necessary inputs: the network and the pairwise distances. Below, we copy the network into a new network, whose edge lengths will be modified during calibration. We first set all edge lengths to some reasonable starting value, e.g.&nbsp;the maximum genetic pairwise distance divided by the number of taxa.</p>
<div class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>net <span class="op">=</span> <span class="fu">deepcopy</span>(net_snaq)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="co"># modify branch lengths for good starting values: 0.03807 max distance / 17 taxa ≈ 0.002</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> <span class="cn">e</span> <span class="kw">in</span> net.edge</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>    <span class="cn">e</span>.length <span class="op">=</span> (<span class="cn">e</span>.hybrid ? <span class="fl">0.0</span> <span class="op">:</span> <span class="fl">0.002</span>)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We are ready to run the calibration. Its optimization criterion is the sum of squares between the observed distances, and the distances expected from the network (weighted average of tree distances, weighted by γ’s). The calibration function modifies the network’s edge lengths directly. Below, we run the calibration 3 times to make sure branch lengths are optimized thoroughly (each one is very fast).</p>
<div class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="im">using</span> <span class="bu">Random; Random</span>.<span class="fu">seed!</span>(<span class="fl">8732</span>) <span class="co"># for reproducibility: you can pick any other seed</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>fmin, xmin, ret <span class="op">=</span> <span class="fu">calibrateFromPairwiseDistances!</span>(net, avD, taxa,</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>        forceMinorLength0<span class="op">=</span><span class="cn">false</span>, NLoptMethod<span class="op">=:</span>LD_MMA);</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>fmin, ret <span class="co"># 0.000575038516459, :FTOL_REACHED</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="10">
<pre><code>(0.00018470183899352304, :FTOL_REACHED)</code></pre>
</div>
</div>
<div class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>fmin, xmin, ret <span class="op">=</span> <span class="fu">calibrateFromPairwiseDistances!</span>(net, avD, taxa,</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>        forceMinorLength0<span class="op">=</span><span class="cn">false</span>, NLoptMethod<span class="op">=:</span>LN_COBYLA);</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>fmin, ret <span class="co"># fmin increased slightly: score got worse</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="11">
<pre><code>(0.00018617396017517004, :MAXEVAL_REACHED)</code></pre>
</div>
</div>
<div class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>fmin, xmin, ret <span class="op">=</span> <span class="fu">calibrateFromPairwiseDistances!</span>(net, avD, taxa,</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>        forceMinorLength0<span class="op">=</span><span class="cn">false</span>, NLoptMethod<span class="op">=:</span>LD_MMA);</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>fmin, ret </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="12">
<pre><code>(0.00018470186946648593, :FTOL_REACHED)</code></pre>
</div>
</div>
<p>Let’s check the estimated edge lengths, in units of substitutions/site for now (as were pairwise distances). We see that there are edges with tiny lengths (e.g.&nbsp;&lt; 1e-5), and a few edges with tiny negative lengths.</p>
<div class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sort!</span>([<span class="cn">e</span>.length for <span class="cn">e</span> <span class="kw">in</span> net.edge])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="13">
<pre><code>41-element Vector{Float64}:
 -1.7904662127921078e-11
  9.884093262680499e-6
  5.34183046749636e-5
  0.00034393799177601624
  0.0007010063133440925
  0.0007844084008373255
  0.0008069860789016378
  0.0008750883231353323
  0.0010021156240143537
  0.0010144667246464924
  0.0012491689774331623
  0.0017106487083121576
  0.001820879461907711
  ⋮
  0.009032745272741213
  0.009336643215198857
  0.00960757482265956
  0.009686943090932907
  0.009686943090932907
  0.009947876264206615
  0.009947876264206615
  0.010011178662144259
  0.01070230088222567
  0.011107456145767353
  0.011107456145767353
  0.018177139899410177</code></pre>
</div>
</div>
<p>Let’s replace any negative edge lengths with 0:</p>
<div class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> <span class="cn">e</span> <span class="kw">in</span> net.edge</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> <span class="cn">e</span>.length <span class="op">&lt;</span> <span class="fl">0</span> <span class="cn">e</span>.length <span class="op">=</span> <span class="fl">0</span>; <span class="cf">end</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>and peek at the resulting network:</p>
<div class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>R<span class="st">"par"</span>(mar<span class="op">=</span>[<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>]);</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>R<span class="st">"par"</span>(cex<span class="op">=</span><span class="fl">0.5</span>) <span class="co"># decrease "character expansion" for smaller annotations later</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>res <span class="op">=</span> <span class="fu">plot</span>(net, useedgelength<span class="op">=</span><span class="cn">true</span>, showedgelength<span class="op">=</span><span class="cn">true</span>);</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-1-netcal0" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="topic1-netcalibration_files/figure-html/fig-1-netcal0-output-1.png" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption">Figure&nbsp;2: Polemonium network: after calibration, edges drawn proportional to their lengths</figcaption><p></p>
</figure>
</div>
</div>
</div>
</section>
<section id="normalize-the-network-height" class="level2">
<h2 class="anchored" data-anchor-id="normalize-the-network-height">normalize the network height</h2>
<p>Optional: Let’s normalize our calibrated network to a height of 1, so edge lengths will be unit-less after. When we will analyze trait evolution later, we will get trait variance rates that represent variance accumulation from the root to the tips, instead of being relative to the molecular rate in substitutions/site.</p>
<p>To normalize our network, we need to length from the root to the tip. <code>getNodeAges</code> will calculate the “age” of each node, <em>assuming</em> that the network is time-consistent (if there are multiple paths from the root to a given node, all these paths have the same lengths) and ultrametric (all tips are at the same distance from the root). It returns the age of nodes in “pre-order”, so the root comes first.</p>
<div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>rootage <span class="op">=</span> <span class="fu">getNodeAges</span>(net)[<span class="fl">1</span>] <span class="co"># 0.018177</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> <span class="cn">e</span> <span class="kw">in</span> net.edge</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>    <span class="cn">e</span>.length <span class="op">/=</span> rootage</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="cf">end</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="fu">getNodeAges</span>(net) <span class="op">|&gt;</span> maximum <span class="co"># 1.0, sanity check</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>R<span class="st">"par"</span>(mar<span class="op">=</span>[<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>]);</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(net, useedgelength<span class="op">=</span><span class="cn">true</span>, xlim<span class="op">=</span>[<span class="fl">1</span>,<span class="fl">2.5</span>], showgamma<span class="op">=</span><span class="cn">true</span>);</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(net, useedgelength<span class="op">=</span><span class="cn">true</span>, xlim<span class="op">=</span>[<span class="fl">1</span>,<span class="fl">2.5</span>], style<span class="op">=:</span>majortree, arrowlen<span class="op">=</span><span class="fl">0.07</span>);</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-1-netcal" class="cell quarto-layout-panel" data-execution_count="15">
<figure class="figure">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-1-netcal-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="topic1-netcalibration_files/figure-html/fig-1-netcal-output-1.png" class="img-fluid figure-img" data-ref-parent="fig-1-netcal"></p>
<p></p><figcaption class="figure-caption">(a) standard plotting type</figcaption><p></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-1-netcal-2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="topic1-netcalibration_files/figure-html/fig-1-netcal-output-2.png" class="img-fluid figure-img" data-ref-parent="fig-1-netcal"></p>
<p></p><figcaption class="figure-caption">(b) ‘major tree’ plotting type</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;3: Polemonium network: after calibration and normalization to height 1</figcaption><p></p>
</figure>
</div>
</div>
<p>Finally, let’s save our calibrated network to a file:</p>
<div class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode julia code-with-copy"><code class="sourceCode julia"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">writeTopology</span>(net,<span class="st">"data/polemonium_network_calibrated.phy"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./topic-overview.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Workshop overview</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./topic2-phyloANOVA.html" class="pagination-link">
        <span class="nav-page-text">Phylogenetic regression</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>